src/mesh/index_derivs.cxx

```diff
@@ -2038,14 +2138,65 @@ const Field3D Mesh::indexVDDY(const Field &v, const Field &f, CELL_LOC outloc, D
       func = lookupUpwindFunc(table, method);
     }
   
-    bindex bx;
-    start_index(&bx);
-    stencil vval, fval;
-    do {
-      f.setYStencil(fval, bx);
+    bool vtest = v.hasYupYdown() && ( (&v.yup() != &v) || (&v.ydown() != &v) );
+    bool ftest = f.hasYupYdown() && ( (&f.yup() != &f) || (&f.ydown() != &f) );
+    if (vtest && ftest) {
+      // f has distinct yup and ydown fields which
+      // will be used to calculate a derivative along
+      // the magnetic field
+      for (const auto &i : result) {
+        // Set stencils
+        stencil fval;
+        fval.c = f[i];
+        fval.p = f.yup()[i.yp()];
+        fval.m = f.ydown()[i.ym()];
+        fval.pp = nan("");
+        fval.mm = nan("");
+
+        result[i] = func(v[i], fval);
+      }
+    } else if (!vtest && !ftest) {
+      // var has no yup/ydown fields, so we need to shift into field-aligned coordinates
       
-      result(bx.jx, bx.jy, bx.jz) = func(v[{bx.jx, bx.jy, bx.jz}], fval);
-    }while(next_index3(&bx));
+      Field3D v_fa = mesh->toFieldAligned(v);
+      Field3D f_fa = mesh->toFieldAligned(f);
+      if (mesh->ystart>1) {
+        // More than one guard cell, so set pp and mm values
+        // This allows higher-order methods to be used
+        for (const auto &i : result) {
+          // Set stencils
+          stencil fval;
+          fval.c = f[i];
+          fval.p = f[i.yp()];
+          fval.m = f[i.ym()];
+          fval.pp = f[i.offset(0,2,0)];
+          fval.mm = f[i.offset(0,-2,0)];
+
+          result[i] = func(v[i], fval);
+        }
+      } else {
+        for (const auto &i : result) {
+          // Set stencils
+          stencil fval;
+          fval.c = f[i];
```

Comment:
setting the stancils seems like it involves copy/paste.
Can this be changed to autogenerated code, or macros?